File: src/main/java/com/manning/simplysend/entities/Order.java
package com.manning.simplysend.entities;

import com.manning.simplysend.exceptions.OrderAlreadyApprovedException;

import javax.persistence.*;
import java.util.List;
import java.util.Objects;

@Entity
@Table(name = "orders")
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Column(name = "status")
    @Enumerated(EnumType.STRING)
    private StatusEnum status;

    @Column(name = "reason")
    private String reason;

    @Column(name = "comment")
    private String comment;

    @OneToOne
    @JoinColumn(name = "user_id", updatable = false)
    private User user;

    @Column(name = "items")
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "order_items", joinColumns = { @JoinColumn(name = "order_id") }, inverseJoinColumns = {
            @JoinColumn(name = "item_id") })
    private List<Item> items;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public StatusEnum getStatus() {
        return status;
    }

    public void setStatus(StatusEnum status) {
        this.status = status;
    }

    public String getReason() {
        return reason;
    }

    public void setReason(String type) {
        this.reason = type;
    }

    public String getComment() {
        return comment;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    public List<Item> getItems() {
        return items;
    }

    public void setItems(List<Item> items) {
        this.items = items;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public void approve() {
        if (getStatus() == Order.StatusEnum.APPROVED) {
            throw new OrderAlreadyApprovedException();
        }
        setStatus(Order.StatusEnum.APPROVED);
    }

    public void reject() {
        if (getStatus() == Order.StatusEnum.DENIED) {
            throw new OrderAlreadyApprovedException();
        }
        setStatus(Order.StatusEnum.DENIED);
    }

    @Override
    public String toString() {
        return "Order{" +
                "id=" + id +
                ", status=" + status +
                ", type='" + reason + '\'' +
                ", comment='" + comment + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (!(o instanceof Order))
            return false;
        Order order = (Order) o;
        return Objects.equals(getId(), order.getId()) &&
                getStatus() == order.getStatus() &&
                Objects.equals(getReason(), order.getReason()) &&
                Objects.equals(getComment(), order.getComment());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getId(), getStatus(), getReason(), getComment());
    }

    public enum StatusEnum {
        REQUESTED("REQUESTED"),
        APPROVED("APPROVED"),
        DENIED("DENIED");

        private String value;

        StatusEnum(String value) {
            this.value = value;
        }

        public static StatusEnum fromValue(String text) {
            for (StatusEnum b : StatusEnum.values()) {
                if (String.valueOf(b.value).equals(text)) {
                    return b;
                }
            }
            return null;
        }

        @Override

        public String toString() {
            return String.valueOf(value);
        }
    }
}

File: src/main/kotlin/com/manning/simplysend/entities/Order.kt
package com.manning.simplysend.entities

import com.manning.simplysend.exceptions.OrderAlreadyApprovedException
import com.manning.simplysend.exceptions.OrderAlreadyDeniedException
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "orders")
class Order {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Long? = null

    @Column(name = "status")
    @Enumerated(EnumType.STRING)
    var status: StatusEnum? = null

    @Column(name = "reason")
    var reason: String? = null

    @Column(name = "comment")
    var comment: String? = null

    @OneToOne
    @JoinColumn(name = "user_id", updatable = false)
    var user: User? = null

    @Column(name = "items")
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "order_items",
        joinColumns = [JoinColumn(name = "order_id")],
        inverseJoinColumns = [JoinColumn(name = "item_id")]
    )
    var items: List<Item>? = null

    fun approve() {
        if (status == StatusEnum.APPROVED) {
            throw OrderAlreadyApprovedException()
        }
        status = StatusEnum.APPROVED
    }

    fun reject() {
        if (status == StatusEnum.DENIED) {
            throw OrderAlreadyDeniedException()
        }
        status = StatusEnum.DENIED
    }

    override fun toString(): String {
        return "Order{" +
                "id=" + id +
                ", status=" + status +
                ", type='" + reason + '\'' +
                ", comment='" + comment + '\'' +
                '}'
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is Order) return false
        val order = o
        return id == order.id && status == order.status && reason == order.reason && comment == order.comment
    }

    override fun hashCode(): Int {
        return Objects.hash(id, status, reason, comment)
    }

    enum class StatusEnum(private val value: String) {
        REQUESTED("REQUESTED"), APPROVED("APPROVED"), DENIED("DENIED");

        override fun toString(): String {
            return value
        }

        companion object {
            fun fromValue(text: String): StatusEnum? {
                for (b in values()) {
                    if (b.value == text) {
                        return b
                    }
                }
                return null
            }
        }
    }
}